### PR TITLE
RWA-5051: Immediate Initiation changes

### PR DESCRIPTION
**Pull request by adityadwadasi**
### Jira link

See [RWA-5051](https://tools.hmcts.net/jira/browse/RWA-5051)

### Security Vulnerability Assessment ###

**CVE Suppression:** Are there any CVEs present in the codebase (either newly introduced or pre-existing) that are being intentionally suppressed or ignored by this commit?
  * [ ] Yes
  * [ ] No

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java [hunk 1]
```diff
@@ -13,13 +28,34 @@ class EventHandlerConfiguration {
...
-        LOG.info("Setting {} state to unconfigured for Task id: {}",
-                 CFT_TASK_STATE_LOCAL_VARIABLE_NAME,
-                 delegateTask.getId());
-        delegateTask.setVariableLocal(CFT_TASK_STATE_LOCAL_VARIABLE_NAME, "unconfigured");
+        if (initiateTasksOnCreate) {
+            LOG.info("Scheduling Task id: {} for initiation after Camunda commit", delegateTask.getId());
+            scheduleInitiateTask(delegateTask);
+        } else {
+            setTaskStateToUnconfigured(delegateTask);
+        }
```
Comment (MartinYSpasov) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java:36]: I'd keep the previous line here for simplicity.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java [hunk 2]
```diff
@@ -34,4 +70,100 @@ public void onTaskDeletedEvent(DelegateTask delegateTask) {
...
+            LOG.warn(
+                "Task id: {} could not be scheduled for initiation because there is no Camunda transaction context.",
+                taskId
+            );
+            setTaskStateToUnconfigured(delegateTask);
+            return;
+        }
+
+        InitiateTaskRequest request = new InitiateTaskRequest(
+            INITIATION_OPERATION,
```
Comment (MartinYSpasov) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: What is the purpose of this static?

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java [hunk 3]
```diff
@@ -34,4 +70,100 @@ public void onTaskDeletedEvent(DelegateTask delegateTask) {
...
+            LOG.warn(
+                "Task id: {} could not be pushed for initiation. Task state will be set to unconfigured.",
+                taskId,
+                ex
+            );
+            setTaskStateToUnconfigured(taskId);
+        }
+    }
+
+    private Map<String, Object> taskAttributes(DelegateTask delegateTask) {
```
Comment (MartinYSpasov) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: Please move this logic to separate class - TaskAttributesMapper

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java [hunk 4]
```diff
@@ -34,4 +70,100 @@ public void onTaskDeletedEvent(DelegateTask delegateTask) {
...
+                taskId,
+                CFT_TASK_STATE_LOCAL_VARIABLE_NAME,
+                UNCONFIGURED_TASK_STATE
+            );
+        } catch (Exception ex) {
+            LOG.error("Task id: {} could not be updated to unconfigured after initiation failure.", taskId, ex);
+        }
+    }
+
+    private void scheduleInitiateTask(DelegateTask delegateTask) {
```
Comment (MartinYSpasov) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: This logic deserves a dedicated  class.

**Review by MartinYSpasov — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java (hunk 1 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java:36]: Made the changes as suggested.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java (hunk 2 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: Made the changes as suggested.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java (hunk 3 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: Made the changes as suggested.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java (hunk 4 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/config/EventHandlerConfiguration.java]: Made the changes as suggested.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+
+    public InitiateTaskRequest create(DelegateTask delegateTask) {
+        return new InitiateTaskRequest(
+            INITIATION_OPERATION,
+            taskAttributes(delegateTask)
+        );
+    }
+
+    private Map<String, Object> taskAttributes(DelegateTask delegateTask) {
```
Comment (MartinYSpasov) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:27]: I'd rename this method to setTaskAttributes() or mapTaskAttributes()

**Review by MartinYSpasov — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        deploymentId = deployment.getId();
+    }
+
+    @AfterEach
+    void tearDown() {
+        repositoryService.deleteDeployment(deploymentId, true);
+    }
+
+    @Test
+    void should_push_task_initiation_request_when_camunda_commit_finished() {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java:127]: Do we have the edge case for if the Camunda commit fails?

File: src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            .singleResult();
+
+        await().atMost(10, SECONDS).untilAsserted(() -> {
+            verify(taskManagementApi).initiateTask(eq(SERVICE_TOKEN), eq(task.getId()), any(InitiateTaskRequest.class));
+            assertThat(taskService.getVariableLocal(task.getId(), CFT_TASK_STATE)).isEqualTo("unassigned");
+        });
+    }
+
+    @Test
+    void should_set_task_state_to_unconfigured_when_feature_toggle_is_enabled_and_push_initiation_fails() {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java]: Is this test name accurate? This is not feature-toggle dependent.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationService.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    private void setTaskStateToUnconfigured(String taskId) {
+        try {
+            taskService.setVariableLocal(
+                taskId,
+                CFT_TASK_STATE_LOCAL_VARIABLE_NAME,
+                "unconfigured"
+            );
+        } catch (Exception ex) {
+            LOG.error("Task id: {} could not be updated to unconfigured after initiation failure.", taskId, ex);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationService.java:39]: Unit test code coverage - edge case for exception not covered.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        Map<String, Object> variables = delegateTask.getVariables();
+        Map<String, Object> attributes = variables == null ? new HashMap<>() : new HashMap<>(variables);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:29]: Partial code coverage. when variable is null need unit test case

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        attributes.remove("assignee");
+        attributes.remove("description");
+        attributes.remove("dueDate");
+        attributes.remove("name");
+        attributes.remove("priorityDate");
+
+        Object taskType = attributes.get("taskType") != null ? attributes.get("taskType") : attributes.get("taskId");
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:37]: Partial code coverage. when variable is null need unit test case

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        putIfNotNull(attributes, "dueDate", format(delegateTask.getDueDate()));
+        putIfNotNull(attributes, "created", format(delegateTask.getCreateTime()));
+        putIfNotNull(attributes, "assignee", delegateTask.getAssignee());
+        putIfNotNull(attributes, "description", delegateTask.getDescription());
+
+        return attributes;
+    }
+
+    private void putIfNotNull(Map<String, Object> attributes, String name, Object value) {
+        if (value != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:50]: Unit test coverage for when value is not null

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            attributes.put(name, value);
+        }
+    }
+
+    private String format(Date date) {
+        if (date == null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:56]: Unit test coverage for when date is not null

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+            return null;
+        }
+        return CAMUNDA_DATA_TIME_FORMATTER.format(date.toInstant().atZone(ZoneId.systemDefault()));
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:59]: Unit test coverage missing for this line

**Review by BenNzewi — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java (hunk 6 above)
Comment (adityadwadasi) [reply src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java:127]: Added an integration test.

File: src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java (hunk 7 above)
Comment (adityadwadasi) [reply src/integrationTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateIntegrationTest.java]: I have renamed the test to reflect the actual behaviour.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationService.java (hunk 8 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationService.java:39]: Added a unit test covering the case where updating cftTaskState to unconfigured fails.

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java (hunk 9 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:29]: Added a unit test

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java (hunk 10 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:37]: Added coverage for an explicitly provided taskType

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java (hunk 11 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:50]: Added a unit test with non-null task

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java (hunk 12 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:56]: Added coverage for non-null creation and due dates

File: src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java (hunk 13 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/camunda/bpm/services/TaskInitiationRequestFactory.java:59]: updated should_prefer_task_type_and_include_non_null_task_metadata()

**Review by larslnde — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+            .pollInterval(2, SECONDS)
+            .atMost(60, SECONDS)
+            .untilAsserted(() -> assertThat(testUtils.cftTaskState(taskId))
+                .isEqualTo("unassigned"));
+    }
+
+    @Test
+    void should_set_task_to_unconfigured_when_camunda_initiation_fails() {
+        CamundaFunctionalTestUtils.ProcessDefinition processDefinition = testUtils.deployTaskProcess();
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: We should add 1 or 2 useful logs to indicate the test is running/has run/has executed?

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test
+    void should_initiate_task_when_task_committed_successfully() {
+        String caseId = testUtils.createWaCcdCase();
+        CamundaFunctionalTestUtils.ProcessDefinition processDefinition = testUtils.deployTaskProcess();
+
+        testUtils.correlateCreateTaskMessage(processDefinition, caseId);
+        taskId = testUtils.getCreatedTaskId(processDefinition.processId());
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: We should add 1 or 2 useful logs to indicate the test is running/has run/has executed?

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java (hunk 14 above)
Comment (adityadwadasi) [reply src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: I updated the tests as discussed.

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java (hunk 15 above)
Comment (adityadwadasi) [reply src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: I updated the tests as discussed.

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        testUtils.setUp();
+    }
+
+    @AfterEach
+    void tearDown() {
+        LOG.info("Cleaning up task initiation functional test resources for task id: {}", taskId);
+        testUtils.cleanUp(taskId);
+        LOG.info("Task initiation functional test cleanup completed");
+    }
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: We should add 1 or 2 useful logs to indicate the test is running/has run/has executed?

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        LOG.info("Created CCD case: {} and deployed Camunda process: {}", caseId, processDefinition.processId());
+
+        testUtils.correlateCreateTaskMessage(processDefinition, caseId);
+        taskId = testUtils.getCreatedTaskId(processDefinition.processId());
+        LOG.info("Created Camunda task: {}. Waiting for cftTaskState to become unassigned", taskId);
+
+        await()
+            .ignoreExceptions()
+            .pollInterval(2, SECONDS)
+            .atMost(60, SECONDS)
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: We should add 1 or 2 useful logs to indicate the test is running/has run/has executed?

File: build.gradle [hunk 18]
```diff
@@ -229,6 +245,7 @@ tasks.withType(Test) {
 
     testLogging {
         exceptionFormat = 'full'
+        showStandardStreams = true
```
Comment (MartinYSpasov) [inline build.gradle]: What is the purpose of this line?

**Review by MartinYSpasov — APPROVED** [review]

File: build.gradle (hunk 18 above)
Comment (adityadwadasi) [reply build.gradle]: The line makes the logs from the tests visible, Martin. This is there in TM and other repo's too.
It was added on Ben's suggestion for seeing logs for passed tests.

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+@SpringBootTest(classes = {
+    CamundaFunctionalTestUtils.class,
+    FunctionalTestServiceAuthConfiguration.class
+})
+@ActiveProfiles("functional")
+@Disabled("Disabled until WA_INITIATE_TASKS_ON_CREATE_ENABLED is enabled")
```
Comment (MartinYSpasov) [inline src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: You cannot disable a full test based on environment variable. This variable may or may not be a true, therefore it must be configurable through the test

**Review by MartinYSpasov — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java (hunk 19 above)
Comment (adityadwadasi) [reply src/functionalTest/java/uk/gov/hmcts/reform/camunda/bpm/config/TaskInitiationOnCreateFunctionalTest.java]: I made changes to ignor the tests programatically.
